examples/cis2-wccd-upgradable/src/lib.rs

```diff
@@ -0,0 +1,3721 @@
+    } = host.state().protocol_addresses
+    {
+        (proxy_address, implementation_address)
+    } else {
+        panic!("{:?}", concordium_cis2::Cis2Error::Custom(CustomContractError::UnInitialized));
+    };
+
+    let state = ReturnBasicState {
+        proxy_address,
+        implementation_address,
+        paused: host.state().paused,
+    };
+    Ok(state)
+}
+
+/// Function to view state of contract.
+#[receive(contract = "CIS2-wCCD", name = "view", return_value = "StateImplementation")]
+fn contract_implementation_view<'a, 'b, S: HasStateApi>(
+    _ctx: &'b impl HasReceiveContext,
+    host: &'a impl HasHost<StateImplementation, StateApiType = S>,
+) -> ContractResult<&'a StateImplementation> {
+    Ok(host.state())
+}
+
+/// Function to view state of contract.
+#[receive(contract = "CIS2-wCCD-Proxy", name = "view", return_value = "StateProxy")]
+fn contract_proxy_view<'a, 'b, S: HasStateApi>(
+    _ctx: &'b impl HasReceiveContext,
+    host: &'a impl HasHost<StateProxy, StateApiType = S>,
+) -> ContractResult<&'a StateProxy> {
+    Ok(host.state())
+}
+
+/// Helper function to ensure contract is not paused.
+fn when_not_paused<S>(
+    host: &mut impl HasHost<StateImplementation, StateApiType = S>,
+) -> ContractResult<()> {
+    let state_address = if let ProtocolAddressesImplementation::Initialized {
+        proxy_address: _,
+        state_address,
+    } = host.state().protocol_addresses
+    {
+        state_address
+    } else {
+        panic!("{:?}", concordium_cis2::Cis2Error::Custom(CustomContractError::UnInitialized));
+    };
+
+    let paused = host.invoke_contract_raw_read_only(
+        &state_address,
+        Parameter(&[]),
+        EntrypointName::new_unchecked("getPaused"),
+        Amount::zero(),
+    )?;
+
+    // It is expected that this contract is initialized with the w_ccd_state
+    // contract (a V1 contract). In that case, the paused variable can be
+    // queried from the state contract without error.
+    let paused: bool = paused
+        .ok_or(concordium_cis2::Cis2Error::Custom(CustomContractError::StateInvokeError))?
+        .get()?;
+    // Check that contract is not paused.
+    ensure!(!paused, concordium_cis2::Cis2Error::Custom(CustomContractError::ContractPaused));
+    Ok(())
+}
+
+/// Wrap an amount of CCD into wCCD tokens and transfer the tokens if the sender
+/// is not the receiver.
+#[receive(contract = "CIS2-wCCD", name = "wrap", mutable, payable)]
+fn contract_wrap<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<StateImplementation, StateApiType = S>,
+    amount: Amount,
+) -> ContractResult<()> {
+    let (proxy_address, state_address) = if let ProtocolAddressesImplementation::Initialized {
+        proxy_address,
+        state_address,
+    } = host.state().protocol_addresses
+    {
+        (proxy_address, state_address)
+    } else {
+        panic!("{:?}", concordium_cis2::Cis2Error::Custom(CustomContractError::UnInitialized));
+    };
```

Comment:
Consider adding a `getProtocolAddresses` method on `StateImplementation`, which returns this pair and bails if not initialized